### ISSUE TITLE
[HOLD for payment 2022-11-02] [$1000] [Form Refactor] AddressForm

### ISSUE THREAD

**Issue by neil-marcellini** [Expensify/App#10738]
Coming from the [New Expensify Forms](https://docs.google.com/document/d/1DI4jQ0nyfi2JczhPT440orwlXsTP_O-aGvg-HLmcbwI/edit#heading=h.mnjguil21u0a) design doc, we should refactor the inputs within the [AddressForm](https://github.com/Expensify/App/blob/68c20364e5bef3fbaf3ce885b4a1d7418ea93e63/src/pages/ReimbursementAccount/AddressForm.js) to be compatible with the new Form component, following the guidelines below:

Here's an example of a Form refactor: https://github.com/Expensify/App/pull/9056
[Here](https://github.com/Expensify/App/blob/8e07c62fb9780c627106f45eb72614466a850f27/src/components/AddressSearch.js#L190) is a component that had its inputs refactored to be compatible with Form.js.

## Guidelines
Contributors should first read [FORMS.md](https://github.com/Expensify/App/blob/b81ff69197e091727b33f8dd53fbc429fdd616b9/contributingGuides/FORMS.md).
The refactored inputs should be functional when wrapped by Form.js and when not.

1. Add an optional inputID prop. 
2. Add an optional shouldSaveDraft prop that defaults to false. 
3. Make the value prop optional.
4. Change any onChange prop, e.g. onTextChange to onChange.
5. In the input’s onBlur method, call props.onBlur().
6. In the input’s onChange method, call props.onChange() .
7. Remove the hasError prop.
8. Update the error message to display if errorText is truthy.
9. Make sure that props.ref is attached to the appropriate DOM node. This could involve forwarding the ref to a child component.
10. Make sure the ref is exposing both value and focus().
11. Add an optional maxLength prop to text inputs and UI to display the current character count out of the limit.
12. Add a hint prop and display the text under the input if there is no existing error.
13. Remove any unused code.

## Testing
Verify that:
- [ ] UI looks as it did before the refactor
- [ ] Values can be added and edited
- [ ] Errors are highlighted correctly (input border)
- [ ] Error messages show up correctly
- [ ] Draft values are saved properly
- [ ] Form alerts are displayed correctly
- [ ] Clicking the `fix the errors` link focuses the first input with error
- [ ] No duplicate submission of the form occurs (when it's already submitting)

**Comment by kakajann** [comment]
May I work on this?
I've refactored AddressForm before in here https://github.com/Expensify/App/pull/7701

**Comment by neil-marcellini** [comment]
I'm going to give priority to @mollfpr coming from the IdentityForm issue. If that doesn't work out for some reason then we will consider your proposal.

**Comment by mollfpr** [comment]
### Proposal

1. Adding `onChange` props to our input component
2. Adding `inputIDs` props to AddressForm with a default value is an empty string for every input. These props will be used to determine what `inputID` is for the street, city, state, and zipCode. 
3. Adding `shouldSaveDraft` props to AddressForm with default `false` and will pass through all the inputs inside AddressForm.
4. Adding `onBlur` props to AddressForm and passing it through the inputs.
5. Change the use of `onChangeText` on TextInput to `onChange`.

```diff
diff --git a/src/components/TextInput/BaseTextInput.js b/src/components/TextInput/BaseTextInput.js
index df829f7fd..e1d5a0cfe 100644
--- a/src/components/TextInput/BaseTextInput.js
+++ b/src/components/TextInput/BaseTextInput.js
@@ -273,6 +273,7 @@ class BaseTextInput extends Component {
                                         maxLength={this.props.maxLength}
                                         onFocus={this.onFocus}
                                         onBlur={this.onBlur}
+                                        onChange={this.props.onChange}
                                         onChangeText={this.setValue}
                                         secureTextEntry={this.state.passwordHidden}
                                         onPressOut={this.props.onPress}
diff --git a/src/components/TextInput/baseTextInputPropTypes.js b/src/components/TextInput/baseTextInputPropTypes.js
index 930f162cf..cc389cedc 100644
--- a/src/components/TextInput/baseTextInputPropTypes.js
+++ b/src/components/TextInput/baseTextInputPropTypes.js
@@ -65,6 +65,9 @@ const propTypes = {
     /** Callback to update the value on Form when input is used in the Form component. */
     onInputChange: PropTypes.func,

+    /** Callback that is called when the text input's text changes */
+    onChange: PropTypes.func,
+
     /** Whether we should wait before focusing the TextInput, useful when using transitions  */
     shouldDelayFocus: PropTypes.bool,
 };
@@ -96,6 +99,7 @@ const defaultProps = {
     hint: '',
     prefixCharacter: '',
     onInputChange: () => {},
+    onChange: () => {},
     shouldDelayFocus: false,
 };

```


```diff
diff --git a/src/pages/ReimbursementAccount/AddressForm.js b/src/pages/ReimbursementAccount/AddressForm.js
index 41c8fa40c..3afbf0bca 100644
--- a/src/pages/ReimbursementAccount/AddressForm.js
+++ b/src/pages/ReimbursementAccount/AddressForm.js
@@ -16,6 +16,8 @@ const propTypes = {
     /** Callback fired when a field changes. Passes args as {[fieldName]: val} */
     onFieldChange: PropTypes.func.isRequired,

+    onBlur: PropTypes.func,
+
     /** Form values */
     values: PropTypes.shape({
         /** Address street field */
@@ -31,6 +33,22 @@ const propTypes = {
         zipCode: PropTypes.string,
     }),

+    inputIDs: PropTypes.shape({
+        /** Address street field */
+        street: PropTypes.string,
+
+        /** Address city field */
+        city: PropTypes.string,
+
+        /** Address state field */
+        state: PropTypes.string,
+
+        /** Address zip code field */
+        zipCode: PropTypes.string,
+    }),
+
+    shouldSaveDraft: PropTypes.bool,
+
     /** Any errors that can arise from form validation */
     errors: PropTypes.objectOf(PropTypes.bool),

@@ -44,44 +62,64 @@ const defaultProps = {
         state: '',
         zipCode: '',
     },
+    inputIDs: {
+        street: '',
+        city: '',
+        state: '',
+        zipCode: '',
+    },
+    shouldSaveDraft: false,
     errors: {},
+    onBlur: () => {},
 };

 const AddressForm = props => (
     <>
         <AddressSearch
+            inputID={props.inputIDs.street}
             label={props.translate(props.streetTranslationKey)}
             containerStyles={[styles.mt4]}
             value={props.values.street}
             onInputChange={props.onFieldChange}
+            onBlur={props.onBlur}
             errorText={props.errors.street ? props.translate('bankAccount.error.addressStreet') : ''}
             hint={props.translate('common.noPO')}
+            shouldSaveDraft={props.shouldSaveDraft}
         />
         <View style={[styles.flexRow, styles.mt4]}>
             <View style={[styles.flex2, styles.mr2]}>
                 <TextInput
+                    inputID={props.inputIDs.city}
                     label={props.translate('common.city')}
                     value={props.values.city}
-                    onChangeText={value => props.onFieldChange({city: value})}
+                    onChange={({nativeEvent}) => props.onFieldChange({city: nativeEvent.text})}
+                    onBlur={props.onBlur}
                     errorText={props.errors.city ? props.translate('bankAccount.error.addressCity') : ''}
+                    shouldSaveDraft={props.shouldSaveDraft}
                 />
             </View>
             <View style={[styles.flex1]}>
                 <StatePicker
+                    inputID={props.inputIDs.state}
                     value={props.values.state}
                     onInputChange={value => props.onFieldChange({state: value})}
+                    onBlur={props.onBlur}
                     errorText={props.errors.state ? props.translate('bankAccount.error.addressState') : ''}
+                    shouldSaveDraft={props.shouldSaveDraft}
                 />
             </View>
         </View>
         <TextInput
+            inputID={props.inputIDs.zipCode}
             label={props.translate('common.zip')}
             containerStyles={[styles.mt4]}
             keyboardType={CONST.KEYBOARD_TYPE.NUMBER_PAD}
             value={props.values.zipCode}
-            onChangeText={value => props.onFieldChange({zipCode: value})}
+            onChange={({nativeEvent}) => props.onFieldChange({zipCode: nativeEvent.text})}
+            onBlur={props.onBlur}
             errorText={props.errors.zipCode ? props.translate('bankAccount.error.zipCode') : ''}
             maxLength={CONST.BANK_ACCOUNT.MAX_LENGTH.ZIP_CODE}
+            shouldSaveDraft={props.shouldSaveDraft}
         />
     </>
 );
```

cc @neil-marcellini

**Comment by mollfpr** [comment]
@neil-marcellini I have checked the `Form.js`, but it can't modify the element inside `AddressForm`. What I understand is in `Form.js` it will loop through the `this.props.children` element and try to find an element with an `inputID` and then modified the prop element. The element nested in `AddressForm` doesn't expose through the element from `Form.js`.

Any idea how `Form.js` can find the input element inside the `AddressForm`? I'm kinda stuck figuring it out.

**Comment by mollfpr** [comment]
I have to refactor `Form.js` to pass the props input inside `AddressForm` from the props (not directly return the element from `Form.js`).

```diff
diff --git a/src/components/Form.js b/src/components/Form.js
index f42101700..70073f9fb 100644
--- a/src/components/Form.js
+++ b/src/components/Form.js
@@ -63,6 +63,7 @@ class Form extends React.Component {
         this.inputValues = {};
         this.touchedInputs = {};

+        this.setPropsForInput = this.setPropsForInput.bind(this);
         this.setTouchedInput = this.setTouchedInput.bind(this);
         this.validate = this.validate.bind(this);
         this.submit = this.submit.bind(this);
@@ -75,6 +76,31 @@ class Form extends React.Component {
         this.touchedInputs[inputID] = true;
     }

+    setPropsForInput(inputID, defaultValue, shouldSaveDraft) {
+        return {
+            ref: node => this.inputRefs[inputID] = node,
+            defaultValue,
+            errorText: this.state.errors[inputID] || '',
+            onBlur: () => {
+                this.setTouchedInput(inputID);
+                this.validate(this.inputValues);
+            },
+            onInputChange: (value, key) => {
+                const inputKey = key || inputID;
+                this.inputValues[inputKey] = value;
+                const inputRef = this.inputRefs[inputKey];
+
+                if (key && inputRef && _.isFunction(inputRef.setNativeProps)) {
+                    inputRef.setNativeProps({value});
+                }
+                if (shouldSaveDraft) {
+                    FormActions.setDraftValues(this.props.formID, {[inputKey]: value});
+                }
+                this.validate(this.inputValues);
+            },
+        };
+    }
+
     submit() {
         // Return early if the form is already submitting to avoid duplicate submission
         if (this.props.formState.isLoading) {
@@ -127,6 +153,26 @@ class Form extends React.Component {
                 return child;
             }

+            if (child.props.inputIDs) {
+                const inputProps = {};
+
+                _.each(child.props.inputIDs, (inputID, key) => {
+                    const defaultValues = child.props.defaultValues || {};
+                    const defaultValue = this.props.draftValues[inputID] || defaultValues[key];
+
+                    // We want to initialize the input value if it's undefined
+                    if (_.isUndefined(this.inputValues[inputID])) {
+                        this.inputValues[inputID] = defaultValue;
+                    }
+
+                    inputProps[key] = this.setPropsForInput(inputID, defaultValue, child.props.shouldSaveDraft);
+                });
+
+                return React.cloneElement(child, {
+                    inputProps,
+                });
+            }
+
             // Depth first traversal of the render tree as the input element is likely to be the last node
             if (child.props.children) {
                 return React.cloneElement(child, {
@@ -149,28 +195,7 @@ class Form extends React.Component {
                 this.inputValues[inputID] = defaultValue;
             }

-            return React.cloneElement(child, {
-                ref: node => this.inputRefs[inputID] = node,
-                defaultValue,
-                errorText: this.state.errors[inputID] || '',
-                onBlur: () => {
-                    this.setTouchedInput(inputID);
-                    this.validate(this.inputValues);
-                },
-                onInputChange: (value, key) => {
-                    const inputKey = key || inputID;
-                    this.inputValues[inputKey] = value;
-                    const inputRef = this.inputRefs[inputKey];
-
-                    if (key && inputRef && _.isFunction(inputRef.setNativeProps)) {
-                        inputRef.setNativeProps({value});
-                    }
-                    if (child.props.shouldSaveDraft) {
-                        FormActions.setDraftValues(this.props.formID, {[inputKey]: value});
-                    }
-                    this.validate(this.inputValues);
-                },
-            });
+            return React.cloneElement(child, this.setPropsForInput(inputID, defaultValue, child.props.shouldSaveDraft));
         });
     }
```


```diff
diff --git a/src/pages/ReimbursementAccount/AddressForm.js b/src/pages/ReimbursementAccount/AddressForm.js
index 41c8fa40c..29110eebd 100644
--- a/src/pages/ReimbursementAccount/AddressForm.js
+++ b/src/pages/ReimbursementAccount/AddressForm.js
@@ -8,13 +8,21 @@ import withLocalize, {withLocalizePropTypes} from '../../components/withLocalize
 import CONST from '../../CONST';
 import StatePicker from '../../components/StatePicker';

+const inputPropsType = PropTypes.shape({
+    ref: PropTypes.func,
+    defaultValue: PropTypes.string,
+    errorText: PropTypes.string,
+    onBlur: PropTypes.func,
+    onInputChange: PropTypes.func,
+});
+
 const propTypes = {

     /** Translate key for Street name */
     streetTranslationKey: PropTypes.string.isRequired,

     /** Callback fired when a field changes. Passes args as {[fieldName]: val} */
-    onFieldChange: PropTypes.func.isRequired,
+    onFieldChange: PropTypes.func,

     /** Form values */
     values: PropTypes.shape({
@@ -34,54 +42,117 @@ const propTypes = {
     /** Any errors that can arise from form validation */
     errors: PropTypes.objectOf(PropTypes.bool),

+    inputIDs: PropTypes.shape({
+        /** Address street field */
+        street: PropTypes.string,
+
+        /** Address city field */
+        city: PropTypes.string,
+
+        /** Address state field */
+        state: PropTypes.string,
+
+        /** Address zip code field */
+        zipCode: PropTypes.string,
+    }),
+
+    inputProps: PropTypes.shape({
+        /** Address street field */
+        street: inputPropsType,
+
+        /** Address city field */
+        city: inputPropsType,
+
+        /** Address state field */
+        state: inputPropsType,
+
+        /** Address zip code field */
+        zipCode: inputPropsType,
+    }),
+
+    shouldSaveDraft: PropTypes.bool,
+
     ...withLocalizePropTypes,
 };

 const defaultProps = {
+    onFieldChange: () => {},
     values: {
         street: '',
         city: '',
         state: '',
         zipCode: '',
     },
+    inputIDs: {
+        street: '',
+        city: '',
+        state: '',
+        zipCode: '',
+    },
+    inputProps: {
+        street: {},
+        city: {},
+        state: {},
+        zipCode: {},
+    },
     errors: {},
+    shouldSaveDraft: false,
 };

 const AddressForm = props => (
     <>
-        <AddressSearch
-            label={props.translate(props.streetTranslationKey)}
-            containerStyles={[styles.mt4]}
-            value={props.values.street}
-            onInputChange={props.onFieldChange}
-            errorText={props.errors.street ? props.translate('bankAccount.error.addressStreet') : ''}
-            hint={props.translate('common.noPO')}
-        />
+        <View>
+            <AddressSearch
+                inputID={props.inputIDs.street}
+                shouldSaveDraft={props.shouldSaveDraft}
+                label={props.translate(props.streetTranslationKey)}
+                containerStyles={[styles.mt4]}
+                value={props.values.street || undefined}
+                onInputChange={props.onFieldChange}
+                errorText={props.errors.street ? props.translate('bankAccount.error.addressStreet') : ''}
+                hint={props.translate('common.noPO')}
+                renamedInputKeys={props.inputIDs}
+                // eslint-disable-next-line react/jsx-props-no-spreading
+                {...props.inputProps.street}
+            />
+        </View>
         <View style={[styles.flexRow, styles.mt4]}>
             <View style={[styles.flex2, styles.mr2]}>
                 <TextInput
+                    inputID={props.inputIDs.city}
+                    shouldSaveDraft={props.shouldSaveDraft}
                     label={props.translate('common.city')}
-                    value={props.values.city}
+                    value={props.values.city || undefined}
                     onChangeText={value => props.onFieldChange({city: value})}
                     errorText={props.errors.city ? props.translate('bankAccount.error.addressCity') : ''}
+                    // eslint-disable-next-line react/jsx-props-no-spreading
+                    {...props.inputProps.city}
                 />
             </View>
             <View style={[styles.flex1]}>
                 <StatePicker
-                    value={props.values.state}
+                    inputID={props.inputIDs.state}
+                    shouldSaveDraft={props.shouldSaveDraft}
+                    value={props.values.state || undefined}
                     onInputChange={value => props.onFieldChange({state: value})}
                     errorText={props.errors.state ? props.translate('bankAccount.error.addressState') : ''}
+                    // eslint-disable-next-line react/jsx-props-no-spreading
+                    {...props.inputProps.state}
                 />
             </View>
         </View>
         <TextInput
+            inputID={props.inputIDs.zipCode}
+            shouldSaveDraft={props.shouldSaveDraft}
             label={props.translate('common.zip')}
             containerStyles={[styles.mt4]}
             keyboardType={CONST.KEYBOARD_TYPE.NUMBER_PAD}
-            value={props.values.zipCode}
+            value={props.values.zipCode || undefined}
             onChangeText={value => props.onFieldChange({zipCode: value})}
             errorText={props.errors.zipCode ? props.translate('bankAccount.error.zipCode') : ''}
             maxLength={CONST.BANK_ACCOUNT.MAX_LENGTH.ZIP_CODE}
+            // eslint-disable-next-line react/jsx-props-no-spreading
+            {...props.inputProps.zipCode}
         />
     </>
 );
```

The usage of `AddressForm` inside `Form.js` will be like this:
```js
                    <AddressForm
                        inputIDs={{
                            street: 'addressStreet',
                            city: 'addressCity',
                            zipCode: 'addressZipCode',
                            state: 'addressState',
                        }}
                        streetTranslationKey="common.companyAddress"
                        shouldSaveDraft
                    />
```

cc @neil-marcellini

**Comment by puneetlath** [comment]
External Upwork job: https://www.upwork.com/jobs/~0145eb26d3c4092442
Internal Upwork job: https://www.upwork.com/ab/applicants/1567094665515032576/job-details

**Comment by bakhtiar120** [comment]
Can i work for this? i have experience create some form use react native. Thank you

**Comment by mollfpr** [comment]
@neil-marcellini @mananjadhav Any thoughts on my proposal https://github.com/Expensify/App/issues/10738#issuecomment-1236347332?

**Comment by mananjadhav** [comment]
Thanks for the bump @mollfpr. Reviewing it currently, and I guess its going to block 2-3 issues. Allow me sometime to write the comment and link the issues.

**Comment by mananjadhav** [comment]
@mollfpr's proposal looks fine to me.  I haven't tested it, but it'll set a good example for us to have subforms being used.

My only question would be: @neil-marcellini Can you confirm if we still want to retain `AddressForm`, and use the way @mollfpr suggested?
```
  <AddressForm
        inputIDs={{
            street: 'addressStreet',
            city: 'addressCity',
            zipCode: 'addressZipCode',
            state: 'addressState',
        }}
        streetTranslationKey="common.companyAddress"
        shouldSaveDraft
    />
```

or should we split `AddressSearch` and other properties are children inside `Form.js` component of the respective form. I am inclined to go with @mollfpr's suggestion.

🎀 👀 🎀 
C+ reviewed

----

Also I am going to put https://github.com/Expensify/App/issues/9580 on hold based on the completion of this issue, based on the [comment here](https://github.com/Expensify/App/pull/10741#issuecomment-1239127226).

**Comment by mollfpr** [comment]
Bump @neil-marcellini 👀

**Comment by puneetlath** [comment]
Ah, this is waiting on me, not @neil-marcellini. I'll take a look shortly.

**Comment by puneetlath** [comment]
Hm, so can we not use the `Form` component inside the `AddressForm` component itself? Like: 

```
AddressForm = props => (
    <AddressSearch />
    <Form>
        <View>
            <Input />
            <Input />
        </View>
    </Form>
)
```

**Comment by mananjadhav** [comment]
Afaik, the usage of `AddressForm` is with other forms like `CompanyStep` that I linked earlier and they would themselves be wrapped around `<Form />` with their own OnyxKeys, and draft state. Having nested `Form` with `AddressForm` won't  work.

**Comment by mollfpr** [comment]
This refactor is a series of refactoring the `RequestorStep` form to use `Form.js`. The `AddressForm` is used on `IdentityForm` and `CompanyStep`.

The structure of `RequestorStep` will be `Form > IdentityForm > AddressForm` and still usable by `CompanyStep > AddressForm` that uses controlled values on the parent component.

Also, `Form.js` cannot modify children inside `AddressForm`, because the children are not accessible.

Is this make sense to you?

cc @puneetlath @mananjadhav

**Comment by luacmartins** [comment]
@mollfpr what @mananjadhav mentioned [here](https://github.com/Expensify/App/issues/10738#issuecomment-1245945285) is correct, we shouldn't nest Form components. I think that we should:

1. Rename AddressForm, since it's just a wrapper around multiple inputs and not a form on its own (it doesn't have a submit button for example)
2. Make sure that the props passed by the parent Form component, e.g. CompanyStep, RequestorStep, etc are used properly

I'd look into the usage of [AddressSearch in AddDebitCardPage](https://github.com/Expensify/App/blob/e8aef52a5f18814bb7729e071df638478426077a/src/pages/settings/Payments/AddDebitCardPage.js#L153-L157) for an example on how to refactor AddressForm. We didn't use AddressForm in AddDebitCardPage because we omit the `city` field, but other address fields are present there and you can see how to handle onChange callbacks.

**Comment by mollfpr** [comment]
@luacmartins Thanks for the feedback!

The problem is props from `Form.js` is not passed to inputs inside `AddressForm`. The children from `AddressForm` is not exposed when doing recursive inside `Form.js`.

**Comment by luacmartins** [comment]
>The problem is props from Form.js is not passed to inputs inside AddressForm. The children from AddressForm is not exposed when doing recursive inside Form.js.

I don't think this is true. The recursive call happens on every child of Form. If you add an inputID to any input in AddressForm, it should receive all props from Form.

**Comment by mollfpr** [comment]
### Proposal

Updating the `Form.js` using `props.type` to expose the children of a component such as `AddressForm` and recursive the children to find an input with `inputID`. But the note is, it only works for a component that does not wrap with HOC.
```diff
diff --git a/src/components/Form.js b/src/components/Form.js
index f42101700..e21290cbb 100644
--- a/src/components/Form.js
+++ b/src/components/Form.js
@@ -134,6 +134,18 @@ class Form extends React.Component {
                 });
             }

+            if (_.isFunction(child.type)) {
+                const nestedChildren = new child.type(child.props);
+
+                if (!React.isValidElement(nestedChildren) || !nestedChildren.props.children) {
+                    return child;
+                }
+
+                return React.cloneElement(nestedChildren, {
+                    children: this.childrenWrapperWithProps(nestedChildren.props.children),
+                });
+            }
+
             // We check if the child has the inputID prop.
             // We don't want to pass form props to non form components, e.g. View, Text, etc
             if (!child.props.inputID) {
```


For the `AddressForm`, 
1. Removing the `withLocalize` to make the traversing children from `Form.js` is working and replace with passed `translate` function from the parent component.
2. Renaming the `onFieldChange` prop to `onInputChange`
3. Adding new prop `renamedInputKeys` and `shouldSaveDraft`. `renamedInputKeys` is used to determine the `inputID` for the input and also passes to `AddressSearch` for auto-fill others input. `shouldSaveDraft` is pass to the input components.
4. Updating the default props for `values` to undefined on each input. For some reason, if the value should be `undefined` for the `Form.js` input works.

>Rename AddressForm, since it's just a wrapper around multiple inputs and not a form on its own (it doesn't have a submit button for example).
Make sure that the props passed by the parent Form component, e.g. CompanyStep, RequestorStep, etc are used properly

The above things are also the changes we need to do for `AddressForm`.


```diff
diff --git a/src/pages/ReimbursementAccount/AddressForm.js b/src/pages/ReimbursementAccount/AddressForm.js
index 41c8fa40c..550a82086 100644
--- a/src/pages/ReimbursementAccount/AddressForm.js
+++ b/src/pages/ReimbursementAccount/AddressForm.js
@@ -4,7 +4,6 @@ import PropTypes from 'prop-types';
 import TextInput from '../../components/TextInput';
 import AddressSearch from '../../components/AddressSearch';
 import styles from '../../styles/styles';
-import withLocalize, {withLocalizePropTypes} from '../../components/withLocalize';
 import CONST from '../../CONST';
 import StatePicker from '../../components/StatePicker';

@@ -13,8 +12,11 @@ const propTypes = {
     /** Translate key for Street name */
     streetTranslationKey: PropTypes.string.isRequired,

+    /** Returns translated string for given locale and phrase */
+    translate: PropTypes.func.isRequired,
+
     /** Callback fired when a field changes. Passes args as {[fieldName]: val} */
-    onFieldChange: PropTypes.func.isRequired,
+    onInputChange: PropTypes.func,

     /** Form values */
     values: PropTypes.shape({
@@ -34,52 +36,80 @@ const propTypes = {
     /** Any errors that can arise from form validation */
     errors: PropTypes.objectOf(PropTypes.bool),

-    ...withLocalizePropTypes,
+    /** A map of inputID key names */
+    renamedInputKeys: PropTypes.shape({
+        street: PropTypes.string,
+        city: PropTypes.string,
+        state: PropTypes.string,
+        zipCode: PropTypes.string,
+    }),
+
+    /** Saves a draft of the input value when used in a form */
+    shouldSaveDraft: PropTypes.bool,
 };

 const defaultProps = {
+    onInputChange: () => {},
     values: {
+        street: undefined,
+        city: undefined,
+        state: undefined,
+        zipCode: undefined,
+    },
+    errors: {},
+    renamedInputKeys: {
         street: '',
         city: '',
         state: '',
         zipCode: '',
     },
-    errors: {},
+    shouldSaveDraft: false,
 };

 const AddressForm = props => (
     <>
-        <AddressSearch
-            label={props.translate(props.streetTranslationKey)}
-            containerStyles={[styles.mt4]}
-            value={props.values.street}
-            onInputChange={props.onFieldChange}
-            errorText={props.errors.street ? props.translate('bankAccount.error.addressStreet') : ''}
-            hint={props.translate('common.noPO')}
-        />
+        <View>
+            <AddressSearch
+                inputID={props.renamedInputKeys.street}
+                shouldSaveDraft={props.shouldSaveDraft}
+                label={props.translate(props.streetTranslationKey)}
+                containerStyles={[styles.mt4]}
+                value={props.values.street}
+                onInputChange={props.onInputChange}
+                errorText={props.errors.street ? props.translate('bankAccount.error.addressStreet') : ''}
+                hint={props.translate('common.noPO')}
+                renamedInputKeys={props.renamedInputKeys}
+            />
+        </View>
         <View style={[styles.flexRow, styles.mt4]}>
             <View style={[styles.flex2, styles.mr2]}>
                 <TextInput
+                    inputID={props.renamedInputKeys.city}
+                    shouldSaveDraft={props.shouldSaveDraft}
                     label={props.translate('common.city')}
                     value={props.values.city}
-                    onChangeText={value => props.onFieldChange({city: value})}
+                    onChangeText={value => props.onInputChange({city: value})}
                     errorText={props.errors.city ? props.translate('bankAccount.error.addressCity') : ''}
                 />
             </View>
             <View style={[styles.flex1]}>
                 <StatePicker
+                    inputID={props.renamedInputKeys.state}
+                    shouldSaveDraft={props.shouldSaveDraft}
                     value={props.values.state}
-                    onInputChange={value => props.onFieldChange({state: value})}
+                    onInputChange={value => props.onInputChange({state: value})}
                     errorText={props.errors.state ? props.translate('bankAccount.error.addressState') : ''}
                 />
             </View>
         </View>
         <TextInput
+            inputID={props.renamedInputKeys.zipCode}
+            shouldSaveDraft={props.shouldSaveDraft}
             label={props.translate('common.zip')}
             containerStyles={[styles.mt4]}
             keyboardType={CONST.KEYBOARD_TYPE.NUMBER_PAD}
             value={props.values.zipCode}
-            onChangeText={value => props.onFieldChange({zipCode: value})}
+            onChangeText={value => props.onInputChange({zipCode: value})}
             errorText={props.errors.zipCode ? props.translate('bankAccount.error.zipCode') : ''}
             maxLength={CONST.BANK_ACCOUNT.MAX_LENGTH.ZIP_CODE}
         />
@@ -89,4 +119,4 @@ const AddressForm = props => (
 AddressForm.propTypes = propTypes;
 AddressForm.defaultProps = defaultProps;
 AddressForm.displayName = 'AddressForm';
-export default withLocalize(AddressForm);
+export default AddressForm;
```


Here's the usage `AddressForm` in `Form.js`.
```js
                    <AddressForm
                        shouldSaveDraft
                        translate={this.props.translate}
                        streetTranslationKey="common.companyAddress"
                        renamedInputKeys={{
                            street: 'addressStreet',
                            city: 'addressCity',
                            state: 'addressState',
                            zipCode: 'addressZipCode',
                        }}
                    />
```

cc @luacmartins @neil-marcellini @mananjadhav

**Comment by mananjadhav** [comment]
I am guessing we can live with `AddressForm` not composed with HoC. But @luacmartins @neil-marcellini can confirm.

**Comment by luacmartins** [comment]
@mananjadhav @mollfpr AddressForm is wrapped with LocaleContext.Consumer, do consumers affect this solution? If not, I think we are ok to not add more HOC around AddressForm.

**Comment by mananjadhav** [comment]
Yes afaik and hence, @mollfpr added `translate` as the prop to `AddressForm`.

**Comment by mollfpr** [comment]
@luacmartins @mananjadhav Are we gonna moving forward with the above proposal?

**Comment by mananjadhav** [comment]
I am fine with the given proposal. Let's hear it from @luacmartins.

**Comment by luacmartins** [comment]
I think the proposal is fine. This will effectively also add support for custom components within Form. I'll let @puneetlath make the final call here, since he's the CME.

**Comment by mananjadhav** [comment]
@puneetlath Quick bump on the above [proposal](https://github.com/Expensify/App/issues/10738#issuecomment-1258335423)

**Comment by puneetlath** [comment]
Sorry for being so slow on this one. I'm good with it to. Let's do it @mollfpr 💪🏾

**Comment by mananjadhav** [comment]
PR is reviewed and approved from my side. Should be close to merge soon.

@puneetlath @luacmartins Quick bump on the PR.

**Comment by puneetlath** [comment]
@mollfpr @mananjadhav the old Upwork job got closed. Can y'all apply to this one so I can get you paid out? Comment here when you have. Thanks!

https://www.upwork.com/jobs/~015ac0b1950f970f4d

**Comment by mananjadhav** [comment]
@puneetlath applied

**Comment by puneetlath** [comment]
All paid. Thanks everyone!
